ggml-alloc.c

```diff
@@ -119,28 +102,20 @@ static void remove_allocated_tensor(struct ggml_allocr * alloc, struct ggml_tens
 }
 #endif
 
-static size_t ggml_allocr_get_alloc_size(struct ggml_allocr * alloc, struct ggml_tensor * tensor) {
-    return ggml_nbytes(tensor);
-
-    UNUSED(alloc);
-}
-
 // check if a tensor is allocated by this buffer
 static bool ggml_allocr_is_own(struct ggml_allocr * alloc, const struct ggml_tensor * tensor) {
-    void * ptr = tensor->data;
-    return ptr >= alloc->data && (char *)ptr < (char *)alloc->data + alloc->max_size;
+    return tensor->buffer == alloc->buffer;
```

Comment:
In train & finetune I make new allocator, allocate tensors, free the allocator and then repeat that for different groups of tensors.
Before this change the actual data pointer was used to check ownership. This pointer differed between successive allocators due to a different data buffer.
But the `buffer` member used here may not differ, because it is malloced by `ggml_backend_buffer_init`. It actually happens that the previous memory address is used, because the previous allocator was freed.

So with this change `ggml_allocr_is_own` does not work correctly with the use case of a temporary allocator.
The allocator must be kept around as long as the tensors itself.

Small example:
```cpp
std::vector<char> buf0;
std::vector<char> buf1;
buf0.resize(...);
buf1.resize(...);
struct ggml_tensor * a = ...;
struct ggml_tensor * b = ...;
struct ggml_allocr * alloc;

alloc = ggml_allocr_new(buf0.data(), buf0.size(), alignment);
ggml_allocr_alloc(alloc, a);
ggml_allocr_free(alloc);

alloc = ggml_allocr_new(buf1.data(), buf1.size(), alignment);
ggml_allocr_alloc(alloc, b);
// Now ggml_allocr_is_own(alloc, a) can be unexpectedly `true`
ggml_allocr_free(alloc);
```

This results in the allocator freeing tensors that are not actually owned by itself.
I write it 'can happen', but I actually observe this behaviour in train & finetune. 
Whether it manifests itself probably depends on malloc.

I can change the train & finetune programs to not free the allocators.
This would solve the problem. 

But in the above example it still is a bit weird, that `ggml_allocr_is_own(alloc, a)` can unexpectedly return `true` when the previous allocator was freed.

Reply:
Maybe I missed other cases, but this should have been fixed in https://github.com/ggerganov/llama.cpp/pull/4486.

The problem is that checking the addresses doesn't really work with measure allocators. The previous implementation used virtual memory as a stopgap, but this has other problems. Ultimately, the issue is that `ggml_allocr_new` creates a temporary `ggml_backend_buffer` that is freed when the allocator is freed, and from that point these tensors cannot be used anymore since they point to a freed buffer. When using `ggml-backend` it is possible to free the allocator if you supply your own buffer, but otherwise the `ggml-alloc` should be kept alive for the lifetime of the tensors.

Reply:
Indeed, this solves issues somewhat for finetune. But there are some more cases. I will push PR after I come back from holiday family visit.